### ISSUE TITLE
[feature for future] Automatic generation of class hierarchy from xml schema

### ISSUE THREAD

**Issue by altavir** [pdvrieze/xmlutil#12]
This feature requires some additional work and external tool, but I am putting it here just to track it. It is possible to generate Kotlin class file compatible with `kotlinx.serialization` from xml schema. We should consider doing it in future. It will also help to cover different use-cases.

**Comment by pdvrieze** [comment]
Of course it would be useful to add the reverse as well (schema from classes). That should be easier to achieve. I'll have to check the state of the code generator (when I have time). The API should support it at this moment, but it was the case that SerialInfo was not implemented completely.

On XML Schema, I'd be happy for some to work, but it is likely that we would be limited to a subset of xml schema. The way it works isn't quite the way that Kotlin/Java type hierarchies work, but supporting a common part of it is certainly possible..

**Comment by pdvrieze** [comment]
The just released version actually builds a full representation of the types. This should make the opposite case (schema from code) quite feasible. That is certainly the easier case, but I don't see why this could not also be used to generate kotlin classes for the schema.

**Comment by altavir** [comment]
Schema from classes is also quite useful.

**Comment by ComBatVision** [comment]
Hi. Any news about this feature release date?

It will be very useful to have alternative to Java XJC + Krasa plugin in Kotlin multi-platform, to have Kotlin classes with validations and text captions from XSD instead of JAXB + JSR303.
If you require, I can provide good complex XSD example.

**Comment by pdvrieze** [comment]
I haven't started on this (work comes first, especially over adding features). Of course a good example would be helpful, but I don't think that I'll be able to put much time in it soon.

**Comment by hfhbd** [comment]
You could take a look at [kfx](https://github.com/hfhbd/kfx) that supports generating Kotlin classes from a WSDL file and additionally XSD files. It's far from perfect but might work for simple classes.
Biggest problem (same as json-schema) is composition and polymorphism.

**Comment by cullub** [comment]
I'll add a +1 for supporting XML schema. Especially serializing from class to xml schema, even just an example policy would be super helpful there.

**Comment by pdvrieze** [comment]
@cullub Do you mean generating schema from the class? That should be doable (I just haven't had time to implement it - schema is quite more complex than ideal)

**Comment by cullub** [comment]
Yep! (And supporting deserialization from schema to the appropriate class type of course)
